### PR TITLE
Test monomorphized AST

### LINKED ISSUE #200: Test monomorphized AST
After monomorphization phase, it will generate a monomorphized AST (MAST). 
To ensure the monomorphization process works as expected, it'd better to have snapshot tests for the generated MASTs for different cases.

### PR DESCRIPTION
**Pull request by cyberbono3**
Closes #200

### CODE REVIEW

**Comment by katat** [comment]
hmm, I think the issue is to add tests to compare the AST during monomorphization phase (MAST). It is useful to review the effects on the generated MAST nodes if there are changes to the logic in MAST pipeline.

There is some tool like https://github.com/mitsuhiko/insta?tab=readme-ov-file to do the job. But I think we should be able to structure the snapshot tests small enough, so the generated AST string can be compared in a manageable way.

**Comment by cyberbono3** [comment]
@katat  Thanks for clarification.  `test_simple_mast` yields some ambiguity in  MAST snapshots. Every new snapshot generates new functions description: `"functions":[["std/bits/nth_bit",` `"functions":[["std/builtins/log"`
You can reproduce it by `cargo insta test --review`

**Comment by katat** [comment]
Ah, it is because the resulted MAST can have the same set of the functions but in different order. The whole MAST contains all the stdlib and builtin functions.

For this kind of test, we should only snapshot the functions in concern, such as the generic examples `test_generic_*`. The snapshot test for MAST is to check if the monomorphization done properly for the generic features.

I think you can create a new test set, similar to the `negative_tests.rs`, for snapshot tests for MAST generation. For the generic feature examples to test snapshot, you can copy the code from the examples folder.

**Comment by cyberbono3** [comment]
> For this kind of test, we should only snapshot the functions in concern, such as the generic examples `test_generic_*`. The snapshot test for MAST is to check if the monomorphization done properly for the generic features.

There are `test_generic_` test cases presented n `src/negative_tests.rs` and in `src/tests/examples.rs`.
I assume, you would like to make snapshot tests for MAST generation  for test cases `test_generic_` listed in  `src/tests/examples.rs` Right?

**Comment by katat** [comment]
> There are test_generic_ test cases presented n src/negative_tests.rs and in src/tests/examples.rs.
I assume, you would like to make snapshot tests for MAST generation for test cases test_generic_ listed in src/tests/examples.rs Right?

Yes, it should only cover the generic functions that involve MAST generation.

> The snapshot test cases from examples folder (arithmetic.no, array.no etc) will produce an ambiguity as I mentioned above.

Not sure If I understood this. The discrepancy is due to the order of the MAST functions. That doesn't mean there is a discrepancy from the MAST. When it comes to comparison, I would suggest checking the monomorphized functions / instantiated function in a narrowed context, that would be easier to find the root causes if there are any discrepancies.

For example, it would be easier to isolate the snapshot test per generic feature, such as generic array:
```
fn clone(const LEN: Field, val: Field) -> [Field; LEN] {
    let mut arr = [0; LEN];
    for idx in 0..LEN {
        arr[idx] = val;
    }
    return arr;
}
```

In other words, in the snapshot test, it should check if the MAST of instantiated function from this generic function is as expected.

File: src/tests/snapshots/mast.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+        let array = [[xx + 1, xx + 2, xx + 3]; 3];
+        let elm = last(array);
+        assert_eq(elm, 4);
+    }
+    "#;
+    test_monomorphization(code);
+}
+
+#[test]
+fn test_generic_builtin_bits() {
```
Comment (cyberbono3) [inline src/tests/snapshots/mast.rs:66]: This tests fails [here](https://github.com/cyberbono3/noname/blob/test-mast/src/mast/mod.rs#L596)

Comment (cyberbono3) [reply src/tests/snapshots/mast.rs:66]: Basically, I have performed snapshot testing for all fields of `Mast(pub TypeChecker). `

**Comment by cyberbono3** [comment]
Basically, I have performed snapshot testing for all fields of `Mast(pub TypeChecker). ` for generic code snippets.

**Comment by cyberbono3** [comment]
@katat  could you take a look , please?
